README.md


# AIAA Propulsive Landing Ground Control

This repository contains scripts and resources to monitor and control the launching of our rocket.

## Organization

 
|Folder|Contents|
|--|--|
|`live_monitoring`|Tools to communicate with rocket during a launch or test|
|`runs`|Data from past launches or tests
|`data_viewer`|Resources to analyze past data
|`structure_manager`|Tools to synchronize ground control with rocket 

## Test / Launch Usage

 1. Install python from [python.org](https://www.python.org/downloads/)
 2. Navigate to the `live_monitoring` folder in your terminal
 3. Run the command `python main.py` (A gui should appear)
 4. Enter the XBee dongle port name into the GUI (e.g. "COM11")
 5. Click the `Select Data Directory` button to select where you want the folder for the run to be generated.
 6. Click the `start recording` button in the GUI
 7. When the test/launch is complete, click `save and exit`
  
  ## Structure Synchronization
  
 1. Run `python structure_manager.py` in the `strucutre_manager` folder
 2. Modify the types, names, and numbers so it matches what is sent from the flight computer
 3. The structure is saved when closing the window

##  Analysis
 - To be completed

## Contributing

Contributions are always welcome!

See `contributing.md` for ways to get started.


history.txt
206 250 186 186 16 91 48 93 102 108 97 115 104 32 115 116 97 114 116 101 100 222 250 254 202 206 250 186 186 24 91 48 93 116 101 108 101 109 101 116 114 121 32 105 110 105 116 105 97 108 105 122 101 100 222 250 254 202 206 250 186 186 16 91 48 93 112 114 111 103 114 97 109 32 115 116 97 114 116 222 250 254 202 206 250 186 186 38 91 48 93 83 80 73 44 32 73 50 67 44 32 82 70 32 99 111 109 109 117 110 105 99 97 116 105 111 110 32 112 105 110 115 32 115 101 116 222 250 254 202 206 250 186 186 18 91 48 93 84 86 67 32 105 110 105 116 105 97 108 105 122 101 100 222 250 254 202 206 250 186 186 17 91 48 93 116 101 115 116 105 110 103 32 84 86 67 46 46 46 222 250 254 202 206 250 186 186 21 91 48 93 84 86 67 32 116 101 115 116 32 99 111 109 112 108 101 116 101 100 222 250 254 202 206 250 186 186 22 91 48 93 115 101 110 115 111 114 115 32 105 110 105 116 105 97 108 105 122 101 100 222 250 254 202 206 250 186 186 36 91 48 93 116 105 109 105 110 103 32 115 101 116 44 32 115 116 97 114 116 105 110 103 32 80 76 95 67 79 85 78 84 68 79 87 78 222 250 254 202 239 190 173 222 0 0 0 0 0 75 92 199 184 226 58 37 64 102 131 76 50 90 188 176 64 0 0 180 251 127 185 17 64 0 0 128 250 255 175 214 63 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 128 35 64 184 30 133 235 81 184 190 63 0 0 0 0 0 0 0 128 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 160 1 36 254 64 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 192 165 119 103 190 0 0 0 96 73 52 137 63 0 0 0 32 97 255 239 63 0 0 0 0 0 0 0 128 0 0 0 32 97 255 239 63 0 0 0 96 73 52 137 191 0 0 0 0 0 0 240 191 0 0 0 192 209 123 2 190 0 0 0 64 49 119 103 190 1 45 145 24 115 52 137 63 0 0 0 160 12 36 46 65 0 0 0 0 0 0 0 0 254 33 20 83 130 160 230 63 78 166 1 197 138 210 113 63 253 33 20 83 130 160 230 63 78 166 1 197 138 210 113 191 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 32 71 244 32 65 222 250 254 202 239 190 173 222 0 30 0 0 0 159 33 28 179 236 161 38 64 44 97 109 140 157 240 82 63 0 0 180 251 127 185 17 64 0 0 128 43 51 195 207 63 92 0 0 0 209 7 0 0 0 0 0 0 93 0 0 0 0 0 0 0 0 0 248 127 0 0 0 0 0 0 0 0 0 0 0 0 0 0 248 127 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 123 20 174 71 225 122 35 64 184 30 133 235 81 184 190 63 0 0 0 0 0 0 0 128 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 160 1 36 254 64 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 53 22 150 217 80 243 195 190 24 195 71 80 129 70 57 191 205 34 226 51 128 221 195 62 30 96 117 98 21 70 57 191 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 42 25 16 218 192 110 16 191 174 216 200 56 73 52 137 63 0 167 83 235 96 255 239 63 0 126 122 58 253 211 242 190 230 154 73 236 96 255 239 63 82 248 55 96 73 52 137 191 86 164 7 189 255 255 239 191 0 14 128 206 181 162 243 190 214 6 235 233 154 95 16 191 246 226 175 215 83 10 218 63 0 0 0 160 12 36 46 65 28 174 97 62 176 64 217 63 118 84 153 233 81 160 230 63 55 14 240 160 78 200 113 63 222 16 16 140 178 160 230 63 174 243 55 219 198 220 113 191 198 193 22 0 109 234 91 191 211 250 41 245 101 103 129 64 13 121 44 172 62 168 95 192 0 0 0 0 0 0 0 0 220 242 70 185 230 31 249 63 0 0 0 0 0 0 0 0 246 226 175 215 83 10 218 63 0 0 0 0 0 0 0 0 220 242 70 185 230 31 249 63 8 88 133 208 119 29 149 192 0 0 0 0 0 0 0 0 246 226 175 215 83 10 218 63 145 249 164 161 104 254 116 192 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 32 71 244 32 65 222 250 254 202 239 190 173 222 0 60 0 0 0 166 123 157 212 151 245 39 64 228 217 229 91 31 214 91 63 0 0 240 46 51 149 17 64 0 0 0 147 153 57 203 63 91 0 0 0 185 7 0 0 0 0 0 0 92 0 0 0 0 0 0 0 0 0 248 127 0 0 0 0 0 0 0 0 0 0 0 0 0 0 248 127 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 128 35 64 41 92 143 194 245 40 188 63 123 20 174 71 225 122 132 63 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 160 1 36 254 64 0 0 0 0 0 0 192 63 0 0 0 0 0 0 176 63 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 204 68 177 86 241 13 73 63 155 20 153 101 98 94 89 191 216 27 40 141 237 254 72 191 199 116 160 52 11 86 89 191 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 15 25 118 150 249 116 6 191 172 18 205 57 57 52 137 63 191 221 175 23 108 255 239 63 92 45 179 120 240 194 100 191 151 69 113 91 101 255 239 63 137 247 210 125 66 52 137 191 170 63 41 22 2 0 240 191 0 231 38 111 164 195 100 191 196 155 68 214 116 105 232 190 138 54 67 39 165 26 249 191 0 0 0 160 12 36 46 65 172 85 167 195 232 246 248 63 110 71 93 177 124 160 230 63 56 165 136 119 229 51 108 63 40 250 102 80 141 160 230 63 117 217 129 85 10 139 117 191 58 121 139 99 253 129 246 191 66 249 209 114 109 103 129 64 165 110 228 185 35 168 95 192 0 0 0 0 0 0 0 0 0 0 0 0 0 0 248 127 0 0 0 0 0 0 0 0 138 54 67 39 165 26 249 191 0 0 0 0 0 0 0 0 0 0 0 0 0 0 248 127 0 0 0 0 0 0 248 255 0 0 0 0 0 0 0 0 138 54 67 39 165 26 249 191 101 197 101 91 110 113 149 64 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 32 71 244 3

rf.py
from multiprocessing import Process, Queue

from data_packet import *
import serial
import tkinter as tk
from struct import *





class RF():
    def __init__(self, port, baud, backlog_threshold = 6000, telem_string='=IhL4d4l26d19d20dI'):
        self._comport = serial.Serial(port, baud)
        self._backlog_bytes_num = backlog_threshold #how many bytes to be in list for a backlog

        self._bytes_received = [] #This list holds recieved bytes and this list is searched through to find packets
        self._history = []
        self._TELEM_HEADER =  [239, 190, 173, 222] #4 byte heaeder to indicate telem frame, 0xDEADBEEF
        self._STRING_HEADER = [206, 250, 186, 186] #4 byte header to indicate string data, 0xBABAFACE
        self._FOOTER = 3405707998 #4 byte uint32_t to indicate ending of string or telem frame // 202, 254, 250, 222

        self._telem_struct = telem_string
        self._sizeofstruct = calcsize(self._telem_struct)

        self._comport.reset_input_buffer()


        self.telem_frame_queue = Queue()
        self.log_queue = Queue()

    def close(self):
        self._comport.close()

    def handle_string_received(self, str):
        self.log_queue.put(str)

    def handle_string_error(self, str):
        self.log_queue.put(str)

    def handle_telem_error(self, error_msg, frame):
        self.log_queue.put(error_msg)
        self.log_queue.put(frame)

    def handle_alignment_notice(self, bytes_skipped):
        self.log_queue.put(str(bytes_skipped) + " required to align")


    #Should be called with high frequency to ensure propper readings
    def read_binary(self):
        if(self._comport.in_waiting >= 5): #waits until there is data of at least the size of the struct because if there is not that much data, there cannot be a complete struct
            
            received = self._comport.read(self._comport.in_waiting) #reads all available data from input buffer into bytearray
            self._bytes_received.extend(list(received)) #adds recieved data to a list
            self._history.extend(list(received))
            
            #print(self._bytes_received)

            if(len(self._bytes_received) > self._backlog_bytes_num):
                #Backlog condition
                self._bytes_received.clear()
                self.log_queue.put("backlog")

            if(len(self._bytes_received) < 10):
                return
            if(self._bytes_received[0:4] != self._TELEM_HEADER and self._bytes_received[0:4] != self._STRING_HEADER): #if the first four bytes received do not match the struct magic number, then alignment must be done
                stra = f'aligned: {self._bytes_received} '
                
                iterations = 0
                while(len(self._bytes_received) >= 4 and (self._bytes_received[0:4] != self._TELEM_HEADER and self._bytes_received[0:4] != self._STRING_HEADER)): #removes from the front of the struct until a head is found
                    self._bytes_received.pop(0)
                    iterations += 1
                stra += f'{iterations} byes'
                self.log_queue.put(stra)

            if(len(self._bytes_received) < 4):
                return

            if((self._bytes_received[0:4] == self._TELEM_HEADER and len(self._bytes_received) >= self._sizeofstruct)): #When there is a valid header and enough bytes are in the list, data is then read.
                res = bytearray(self._bytes_received[0:self._sizeofstruct])
                del self._bytes_received[0:self._sizeofstruct]
                frame = unpack(self._telem_struct, res)
                
                
                if(frame[-1:][0] != self._FOOTER):
                    self.handle_telem_error("invalid frame", frame)
                else:
                    self.telem_frame_queue.put(frame)
                    

            if(self._bytes_received[0:4] == self._STRING_HEADER and len(self._bytes_received) >= 9):
                byte_after_header = bytearray(self._bytes_received[4:5]) #first four bytes are header, 5th byte is size
                length = unpack('=B', byte_after_header)[0]

                if(len(self._bytes_received[5:]) >= length + 5):
                    incomingString = bytearray(self._bytes_received[5:5+length+4]) #length of footer is 4
                    del self._bytes_received[0:5+length+4]
                    string_with_struct = '=' + str(length) + 'cI'
                    parsedString = unpack(string_with_struct, incomingString)
                    
                    if(parsedString[-1:][0] != self._FOOTER):
                        self.handle_string_error("Invalid String, no footer" + str(parsedString))
                    else:
                        try:
                            byte_chars = [s.decode() for s in parsedString[:-1]]
                            output_string = "".join(byte_chars)
                            self.handle_string_received(output_string)
                        except:
                            self.handle_string_error("Character value exceeds ascii values" + str(parsedString))

    def read_string(self):
        line = self._comport.readline()
        arr = line.split(',')
        if(arr[0] == "DATA"):
            self.telem_frame_queue.put(tuple(arr[1:]))
        else:
            self.log_queue.put(arr[1])

    def arm(self):
        self._comport.write(bytes('arm\n', 'utf8'))

    def abort(self):
        self._comport.write(bytes('abort\n', 'utf8'))

    def is_open(self):
        return self._comport.isOpen()